design/crypto_primitives.rs

```diff
@@ -0,0 +1,171 @@
+//! cases, we don't need signing key at all, e.g. in Crust. So this document proposes to split
+//! `PublicKeys` and `SecretKeys` into corresponding types: `PublicEncryptKey, `SecretEncryptKey`,
+//! `PublicSignKey` and `SecretSignKey`.
+
+// whatever is defined in the underlying crypto library
+const PUBLIC_ENCRYPT_KEY_BYTES: usize = rust_sodium::crypto::box_::PUBLICKEYBYTES;
+const SECRET_ENCRYPT_KEY_BYTES: usize = rust_sodium::crypto::sign::SECRETKEYBYTES;
+const PUBLIC_SIGN_KEY_BYTES: usize = rust_sodium::crypto::sign::PUBLICKEYBYTES;
+const SECRET_SIGN_KEY_BYTES: usize = rust_sodium::crypto::sign::SECRETKEYBYTES;
+const SHARED_SECRET_KEY_BYTES: usize = XXX;
+const SIGNATURE_BYTES: usize = XXX;
+const SYMMETRIC_KEY_BYTES: usize = XXX;
+
+quick_error! {
+    /// This error is returned if encryption or decryption fail.
+    /// The encryption failure is rare and mostly connected to serialisation failures.
+    /// Decryption can fail because of invalid keys, invalid data, or deserialisation failures.
+    #[derive(Debug)]
+    pub enum Error {
+        /// Occurs when serialisation or deserialisation fails.
+        Serialisation(e: SerialisationError) {
+            display("error serialising or deserialising message: {}", e)
+            cause(e)
+            from()
+        }
+        /// Occurs when we can't decrypt a message or verify the signature.
+        DecryptVerify(_e: ()) {
+            display("error decrypting/verifying message")
+            from()
+        }
+        /// Occurs in case of an error during initialisation.
+        InitError(e: i32) {
+            display("error while initialising random generator")
+            from()
+        }
+        /// Occurs when we fail to derive encryption key from password.
+        DeriveKey {
+            display("error deriving encryption key from password")
+        }
+    }
+}
+
+/// Derives encryption key from plaintext password.
+pub fn key_from_password(password: &[u8], salt: &[u8]) -> Result<SymmetricKey, Error>;
+
+/// Use `encrypt_key_pair()` to generate public and private keys pair.
+// Make it `Copy`able?
+#[derive(Debug, Hash, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize, Clone)]
+pub struct PublicEncryptKey {
+    encrypt: box_::PublicKey,
+}
+
+impl PublicEncryptKey {
+    /// Construct public key from bytes. Useful when it was serialized before.
+    pub fn from_bytes(public_key: [u8; PUBLIC_ENCRYPT_KEY_BYTES]) -> Self
+
+    /// For anyone who wants to store public key.
+    pub fn as_bytes(self) -> [u8; PUBLIC_ENCRYPT_KEY_BYTES]
+
+    pub fn encrypt_anonymous<T: Serialize>(&self, plaintext: &T) -> Result<Vec<u8>, Error>
+
+    pub fn encrypt_anonymous_bytes(&self, plaintext: &[u8]) -> Vec<u8> {
+}
+
+/// Store secret key in `Arc` to discourage cloning it. Makes it more secure by not duplicating
+/// encryption keys in memory.
+/// Use `encrypt_key_pair()` to generate public and private keys pair.
+#[derive(Debug, PartialEq, Eq, Clone)]
```

Comment:
From what I understood, we will need to also be `Serialize` and `Deserialize` as it's needed for self_authentication (we want to store some encrypted secrets on the network).
Applies for `SecretEncryptKey` (line 87) and `SecretEncryptKeyInner` (line 92)

Reply:
Do we really want to use `Serialize/Deserialize` as they add some metadata or `as_bytes()/from_bytes()` instead? E.g. in case of signatures we want to attach just a signature without `serde` metadata, hence we use `as_bytes()`. Not sure about secret key serialization though.

Reply:
I don't see a strong argument against `as_bytes()`, `from_bytes()`, but maybe @ustulation has more insight on this?